### PR TITLE
Safe call operator

### PR DESCRIPTION
**Pull request by privat**
A long time ago, there was a proposal to have a safe call operator in Nit #1274 `x?.foo` that executes the call if `x` is not null and returns null otherwise (instead of aborting).

This was refused because at the time, the syntax `x?.foo` was considered weird and not POLA.
Moreover, what was proposed was a more general version of the concept that could be used everywhere, not only as a receiver that made the semantic quite complex and even less POLA.

Nowadays, most languages have adopted it https://en.wikipedia.org/wiki/Safe_navigation_operator so the syntax and behavior is not as weird as before.

This operator is nice in the context of Nit with nullable types since if avoids to use chained `if x != null then x.foo` and plays nicely with the type system.

Currently, only explicit dotted calls are handled; e.g. `x?.foo`, I'm not sure if something should be done with other call constructions like `x + y` or `x[y]`.

### CODE REVIEW

File: src/interpreter/naive_interpreter.nit [hunk 1]
```diff
@@ -721,6 +721,10 @@ abstract class Instance
 	# Abort if the instance is not a boolean value.
 	fun is_true: Bool do abort
 
+	# Return `true` is the instance is null.
```
Comment (R4PaSs) [inline src/interpreter/naive_interpreter.nit]: s/is the instance/if the instance

File: src/interpreter/naive_interpreter.nit [hunk 2]
```diff
@@ -721,6 +721,10 @@ abstract class Instance
...
+	# Return `false` otehrwise.
```
Comment (R4PaSs) [inline src/interpreter/naive_interpreter.nit]: s/otehrwise/otherwise

File: src/parser/parser_prod.nit [hunk 3]
```diff
@@ -7271,6 +7271,48 @@ redef class ANamedargExpr
...
+			return
+		end
+	end
+
+	redef fun n_expr=(node)
+	do
+		_n_expr = node
+		node.parent = self
+	end
+	redef fun n_quest=(node)
```
Comment (Morriar) [inline src/parser/parser_prod.nit]: Why no new line before but two after?

File: src/semantize/typing.nit [hunk 4]
```diff
@@ -1915,6 +1919,10 @@ redef class ASendExpr
 	do
 		var nrecv = self.n_expr
 		var recvtype = v.visit_expr(nrecv)
+		if nrecv isa ASafeExpr then
```
Comment (Morriar) [inline src/semantize/typing.nit]:
```nit
is_safe = nrecv isa ASafeExpr
```
?

File: src/semantize/typing.nit [hunk 5]
```diff
@@ -2448,6 +2460,28 @@ redef class AIssetAttrExpr
...
+redef class ASafeExpr
+	redef fun accept_typing(v)
+	do
+		var mtype = v.visit_expr(n_expr)
+		if mtype == null then return # Skip error
+
+		self.mtype = mtype
+
+		if mtype isa MNullType then
+			v.modelbuilder.warning(self, "useless-safe", "Warning: safe operator useless on null.")
```
Comment (Morriar) [inline src/semantize/typing.nit]:
```
Error: can't use safe operator on null?
```

If statically this thing is `null` and you're using a safe operator on it, you're doing something wrong?

File: src/semantize/typing.nit [hunk 6]
```diff
@@ -2448,6 +2460,28 @@ redef class AIssetAttrExpr
...
+			return
+		end
+
+		if not v.can_be_null(mtype) then
+			v.modelbuilder.warning(self, "useless-safe", "Warning: safe operator useless on non nullable.")
```
Comment (Morriar) [inline src/semantize/typing.nit]: Better wording:
```
Warning: useless safe operator on non-nullable value.
```

File: src/parser/parser_prod.nit (hunk 3 above)
Comment (privat) [reply src/parser/parser_prod.nit]: parser_prod.nit is generated. Git metadata is set on this file to disable the diff but github does not care and shows it anyway :(

File: src/semantize/typing.nit (hunk 5 above)
Comment (privat) [reply src/semantize/typing.nit]: yeah, static calls on null suck. They are too weird to be correct but they have a clear unambiguous semantic so if feel extra pedant to forbid them. They could also naturally occur in dead code or in generated code, especially with type adaptation.

Nevertheless, currently `null.foo` is forbidden. It's coherent to forbid `null?.foo` too I assume.

File: src/semantize/typing.nit (hunk 4 above)
Comment (privat) [reply src/semantize/typing.nit]: I will add a comment

File: src/semantize/typing.nit [hunk 7]
```diff
@@ -1903,6 +1903,10 @@ redef class ASendExpr
 	# The property invoked by the send.
 	var callsite: nullable CallSite
 
+	# Is self a safe call (with x?.foo)
+	# If yes and the receiver is null, then the arguments wont be evaluated and the call skipped (replaced with null).
```
Comment (R4PaSs) [inline src/semantize/typing.nit]: `s/If yes/If so`

File: src/semantize/typing.nit (hunk 4 above)
Comment (privat) [reply src/semantize/typing.nit]: comment added

**Comment by privat** [comment]
reroll: applied suggestions and added comments

**Pull request merged**